File: Objective-C/Reference App/18. Display Driver Motion/DisplayDriverMotionViewController.m
//
//  DisplayDriverMotionViewController.m
//  Reference App
//
//  Created by David Deller on 5/29/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

#import "DisplayDriverMotionViewController.h"

@import MapKit;

@interface DisplayDriverMotionViewController ()

@property (weak, nonatomic) IBOutlet TGMapView *mapView;

@property (nonatomic, nullable) JFRWebSocket *socket;

// Put these in your @interface
@property (nonatomic, nonnull) MGLPointAnnotation *driverAnnotation;
@property (nonatomic, nullable) UIViewPropertyAnimator *animator;

@end

@implementation DisplayDriverMotionViewController

typedef NSString *DriverEventType NS_TYPED_ENUM;

DriverEventType const DriverEventTypeMotion = @"motion";

// Make sure to init the annotation when the view controller class loads
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.driverAnnotation = MGLPointAnnotation.new;
    self.mapView.delegate = self;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self setupWebSocket];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self teardownWebSocket];
}

- (void)setupWebSocket {
    JFRWebSocket *socket = [JFRWebSocket.alloc initWithURL:[NSURL URLWithString:@"ws://localhost:3200/"] protocols:@[]];
    
    socket.onConnect = ^{
        NSLog(@"WebSocket is connected");
    };
    
    socket.onDisconnect = ^(NSError *error) {
        NSLog(@"WebSocket is disconnected: %@", error.userInfo);
    };
    
    socket.onText = ^(NSString *text) {
        NSError *error = nil;
        NSDictionary *data = [NSJSONSerialization JSONObjectWithData:[text dataUsingEncoding:NSUTF8StringEncoding] options:0 error:&error];
        if (error != nil || ![data isKindOfClass:NSDictionary.class]) {
            NSLog(@"Error parsing JSON from WebSocket: %@, JSON: %@", error.userInfo, text);
            return;
        }
        
        DriverEventType eventType = data[@"event_type"];
        NSDictionary *payload = data[@"payload"];
        
        if (![eventType isKindOfClass:NSString.class] || ![payload isKindOfClass:NSDictionary.class]) {
            NSLog(@"JSON not in expected format: %@", text);
            return;
        }
        
        [self handleEvent:eventType payload:payload];
    };
    
    [socket connect];
    
    self.socket = socket;
}

- (void)teardownWebSocket {
    [self.socket disconnect];
}

- (void)handleEvent:(nonnull DriverEventType)eventType payload:(nonnull NSDictionary *)payload {
    if ([eventType isEqualToString:DriverEventTypeMotion]) {
        NSArray<NSArray<NSNumber *> *> *points = payload[@"points"];
        NSNumber *timeIntervalObj = payload[@"timeInterval"];
        if (![points isKindOfClass:NSArray.class] || ![points.firstObject isKindOfClass:NSArray.class] || ![timeIntervalObj isKindOfClass:NSNumber.class]) {
            NSLog(@"Payload not in expected format: %@", payload);
            return;
        }
        
        NSTimeInterval timeInterval = timeIntervalObj.doubleValue;
        
        NSMutableArray<NSValue *> *locations = [NSMutableArray arrayWithCapacity:points.count];
        for (NSArray<NSNumber *> *coordinateArray in points) {
            NSNumber *latitudeObj = coordinateArray[0];
            NSNumber *longitudeObj = coordinateArray[1];
            if (![latitudeObj isKindOfClass:NSNumber.class] || ![longitudeObj isKindOfClass:NSNumber.class]) {
                NSLog(@"Payload not in expected format: %@", payload);
                return;
            }
            
            CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake(latitudeObj.doubleValue, longitudeObj.doubleValue);
            [locations addObject:[NSValue valueWithMKCoordinate:coordinate]];
        }
        
        [self handleMotionEventWithLocations:[NSArray arrayWithArray:locations] timeInterval:timeInterval];
        
    } else {
        NSLog(@"Unexpected event type: %@", eventType);
    }
}

- (void)handleMotionEventWithLocations:(NSArray<NSValue *> *)locations timeInterval:(NSTimeInterval)timeInterval {
    // We need at least one location
    if (locations.firstObject == nil) {
        return;
    }
    CLLocationCoordinate2D firstLocation = locations.firstObject.MKCoordinateValue;
    
    // Stop any previous animation that may still be running
    [self.animator stopAnimation:YES];
    self.animator = nil;
    
    // Add the annotation to the map, if it hasn't been already
    if (self.mapView.annotations == nil || ![self.mapView.annotations containsObject:self.driverAnnotation]) {
        [self.mapView addAnnotation:self.driverAnnotation];
    }
    
    // Zoom/pan the map to fit all of the coordinates, so the marker won't go off-screen
    NSInteger numCoordinates = locations.count;
    CLLocationCoordinate2D coordinatesCArray[numCoordinates];
    NSInteger index = 0;
    for (NSValue *locationValue in locations) {
        CLLocationCoordinate2D coordinate = locationValue.MKCoordinateValue;
        coordinatesCArray[index] = coordinate;
        index += 1;
    }
    [self.mapView setVisibleCoordinates:coordinatesCArray count:numCoordinates edgePadding:UIEdgeInsetsMake(30, 30, 30, 30) animated:NO];
    
    // Start out by jumping immediately to the first coordinate
    self.driverAnnotation.coordinate = firstLocation;
    
    NSArray<NSValue *> *locationsAfterFirst = [locations subarrayWithRange:NSMakeRange(1, (locations.count - 1))];
    
    // We need more coordinates if we're going to animate them
    if (locationsAfterFirst.count <= 0) {
        return;
    }
    
    UIViewPropertyAnimator *animator = [UIViewPropertyAnimator.alloc initWithDuration:timeInterval timingParameters:[UICubicTimingParameters.alloc initWithAnimationCurve:UIViewAnimationCurveLinear]];
    double eachKeyframePercent = (timeInterval / locationsAfterFirst.count);
    index = 0;
    
    // Loop through each of the coordinates (besides the first, which we already jumped to) and animate each of them in sequence
    for (NSValue *locationValue in locationsAfterFirst) {
        [animator addAnimations:^{
            self.driverAnnotation.coordinate = locationValue.MKCoordinateValue;
            
        } delayFactor:(index * eachKeyframePercent)];
        
        index += 1;
    }
    
    // Start the animation
    [animator startAnimation];
    
    // Save the animator for later, in case we need to interrupt it when a new one is ready
    self.animator = animator;
}

#pragma mark - MGLMapViewDelegate

- (MGLAnnotationView *)mapView:(MGLMapView *)mapView viewForAnnotation:(nonnull id<MGLAnnotation>)annotation {
    static NSString *ID = @"driver annotation view";
    MGLAnnotationView *annotationView = [mapView dequeueReusableAnnotationViewWithIdentifier:ID];
    if (annotationView != nil) {
        return annotationView;
    } else {
        MGLAnnotationView *view = [MGLAnnotationView.alloc initWithAnnotation:annotation reuseIdentifier:ID];
        
        UIImageView *imageView = [UIImageView.alloc initWithImage:TallyGoStyleKit.imageOfGenericPlaceIcon];
        view.bounds = imageView.frame;
        [view addSubview:imageView];
        
        return view;
    }
}

@end

File: Swift 4/Reference App/18. Display Driver Motion/DisplayDriverMotionViewController.swift
//
//  DisplayDriverMotionViewController.swift
//  Reference App
//
//  Created by David Deller on 5/25/18.
//  Copyright © 2018 TallyGo. All rights reserved.
//

import UIKit
import TallyGoKit
import Starscream

class DisplayDriverMotionViewController: ExampleViewController, MGLMapViewDelegate {
    
    enum EventType: String {
        case motion = "motion"
    }

    @IBOutlet weak var mapView: TGMapView!
    
    // MARK: -

    override func viewDidLoad() {
        super.viewDidLoad()

        mapView.delegate = self
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        setupWebSocket()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        teardownWebSocket()
    }

    var socket: WebSocket?
    
    func setupWebSocket() {
        let socket = WebSocket(url: URL(string: "ws://localhost:3200/")!)
        
        socket.onConnect = {
            print("WebSocket is connected")
        }
        
        socket.onDisconnect = { (error: Error?) in
            print("WebSocket is disconnected: \(error?.localizedDescription ?? "")")
        }
        
        socket.onText = { (text: String) in
            do {
                var data = try JSONSerialization.jsonObject(with: text.data(using: .utf8)!, options: []) as! Dictionary<String, Any>
                
                guard let eventTypeString = data["event_type"] as? String,
                    let eventType = EventType(rawValue: eventTypeString),
                    let payload = data["payload"] as? Dictionary<String, Any> else {
                        
                        NSLog("JSON not in expected format: \(text)")
                        return
                }
                
                self.handleEvent(eventType: eventType, payload: payload)
                
            } catch (let error) {
                NSLog("Error parsing JSON from WebSocket: \(error), JSON: \(text)")
            }
        }
        
        socket.connect()
        
        self.socket = socket
    }

    func teardownWebSocket() {
        socket?.disconnect()
    }

    func handleEvent(eventType: EventType, payload: Dictionary<String, Any>) {
        switch eventType {
        case .motion:
            guard let points = payload["points"] as? [[Double]], let timeInterval = payload["timeInterval"] as? TimeInterval else {
                NSLog("Payload not in expected format: \(payload)")
                return
            }
            
            let locations = points.compactMap { (coordinatePair) -> CLLocationCoordinate2D? in
                if let latitude = coordinatePair.first, let longitude = coordinatePair.last {
                    return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
                } else {
                    return nil
                }
            }
            
            self.handleMotionEvent(locations: locations, timeInterval: timeInterval)
        }
    }
    
    var driverAnnotation: MGLPointAnnotation = MGLPointAnnotation()
    var animator: UIViewPropertyAnimator?
    
    func handleMotionEvent(locations: [CLLocationCoordinate2D], timeInterval: TimeInterval) {
        // We need at least one location
        guard let firstLocation = locations.first else {
            return
        }
        
        // Stop any previous animation that may still be running
        self.animator?.stopAnimation(true)
        self.animator = nil
        
        // Add the annotation to the map, if it hasn't been already
        if mapView.annotations == nil || !(mapView.annotations! as NSArray).contains(driverAnnotation) {
            mapView.addAnnotation(driverAnnotation)
        }
        
        // Zoom/pan the map to fit all of the coordinates, so the marker won't go off-screen
        mapView.setVisibleCoordinates(locations, count: UInt(locations.count), edgePadding: UIEdgeInsets(top: 30, left: 30, bottom: 30, right: 30), animated: false)
        
        // Start out by jumping immediately to the first coordinate
        driverAnnotation.coordinate = firstLocation
        
        var locationsAfterFirst = locations
        locationsAfterFirst.removeFirst()
        
        // We need more coordinates if we're going to animate them
        guard locationsAfterFirst.count > 0 else {
            return
        }
        
        let animator = UIViewPropertyAnimator(duration: timeInterval, timingParameters: UICubicTimingParameters(animationCurve: .linear))
        let eachKeyframePercent: Double = (timeInterval / Double(locationsAfterFirst.count))
        var index = 0
        
        // Loop through each of the coordinates (besides the first, which we already jumped to) and animate each of them in sequence
        locationsAfterFirst.forEach { (location) in
            animator.addAnimations({
                self.driverAnnotation.coordinate = location
                
            }, delayFactor: (CGFloat(index) * CGFloat(eachKeyframePercent)))
            
            index += 1
        }
        
        // Start the animation
        animator.startAnimation()
        
        // Save the animator for later, in case we need to interrupt it when a new one is ready
        self.animator = animator
    }

    // MARK: - MGLMapViewDelegate
    
    func mapView(_ mapView: MGLMapView, viewFor annotation: MGLAnnotation) -> MGLAnnotationView? {
        let ID = "driver annotation view"
        if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: ID) {
            return annotationView
        } else {
            let view = MGLAnnotationView(annotation: annotation, reuseIdentifier: ID)
            
            let imageView = UIImageView(image: TallyGoStyleKit.imageOfGenericPlaceIcon()!)
            view.bounds = imageView.frame
            view.addSubview(imageView)
            
            return view
        }
    }

}

